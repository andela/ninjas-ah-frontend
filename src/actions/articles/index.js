import { createPost } from './createPost';
import { editPost } from './editPost';
import { fetchOneArticle } from './fetchOneArticle';
import { getAllArticles } from './getAllArticles';
import { deleteArticle } from './deleteArticle';
import { publishArticle } from './publishArticle';
import { unpublishArticle } from './unpublishArticle';
import { getPublished } from './getPublished';

export {
  createPost,
  editPost,
  deleteArticle,
  publishArticle,
  unpublishArticle,
  getAllArticles,
  fetchOneArticle,
  getPublished
};
