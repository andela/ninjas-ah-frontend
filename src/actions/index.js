import * as userAction from './user';
import * as notificationActions from './notificationActions';
import {
  getAllArticles,
  fetchOneArticle,
  createPost,
  editPost,
  deleteArticle,
  publishArticle,
  unpublishArticle,
  getPublished
} from './articles';
import { createRate, clearCreateRateStore } from './rating';
import { uploadImage } from './images';
import searchArticles from './searchArticles';

export {
  userAction,
  notificationActions,
  getAllArticles,
  fetchOneArticle,
  createPost,
  editPost,
  deleteArticle,
  publishArticle,
  unpublishArticle,
  uploadImage,
  getPublished,
  createRate,
  clearCreateRateStore,
  searchArticles
};
