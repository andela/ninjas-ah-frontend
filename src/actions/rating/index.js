import { createRate, clearCreateRateStore } from './createRate';

export { createRate, clearCreateRateStore };
