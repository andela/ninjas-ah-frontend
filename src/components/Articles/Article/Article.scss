@import '../../../assets/css/_padding';
@import '../../../assets/css/_radius';
@import '../../../assets/css/_colors';
@import '../../../assets/css/_shadow';

#article {
  .modal-content {
    background: initial;
    border: 0;
    box-shadow: none;
    .modal-body {
      @extend .white, .radius-2;
      padding: 5px;
    }
  }
  .articleBody {
    line-height: 1.8;
    max-width: 100%;
    ::selection {
      font-weight: bold !important;
      background: lightgreen !important;
    }
    .selection {
      background: lightgreen !important;
      font-style: italic;
      font-weight: bold;
    }

    span[style*='--highlight-id'] {
      background-color: lightgreen;
      cursor: pointer;
      col &:hover {
        opacity: 0.5;
      }
    }
  }
  .articleBody img {
    max-width: 100%;
    margin: 10px auto;
    display: block;
  }
  .articleBody img {
    padding: $small/2;
    background: $light !important;
    @extend .shadow-2;
  }
  .articleBody strong {
    font-weight: 800 !important;
  }
  .articleBody blockquote {
    padding: 20px;
    background: #efefef;
    border-left: 5px solid #343434;
  }
  .articleBody a {
    color: #0b3b94 !important;
  }
  .articleInfo {
    @extend .xxlarge-v-padding;
    .avatar {
      position: relative;
      font-size: ($small * 4/3);
      float: left;
    }
    .avatar img {
      position: absolute;
      top: -8px;
      left: -10px;
      width: 32px;
      height: 32px;
      overflow: hidden;
      border-radius: 50%;
    }
    .avatar span {
      position: relative;
      margin-left: 30px;
      margin-right: 15px;
      top: 0px;
    }
    .small-button {
      padding: $small/3 $small !important;
      @extend .radius-4;
    }
    .rating {
      float: right;
    }
  }

  .tags-values {
    float: left;
    margin: 15px 15px 15px 0;
  }

  .value-tag {
    color: $black;
    background: $lighten-grey;
    padding: 9px;
    border-radius: 16px;
    -webkit-border-radius: 16px;
    -moz-border-radius: 16px;
    -ms-border-radius: 16px;
    -o-border-radius: 16px;
  }
}
