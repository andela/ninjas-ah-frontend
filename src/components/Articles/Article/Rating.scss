@import '../../../assets/css/_pixels';
@import '../../../assets/css/_colors';
@import '../../../assets/css/_border';
@import '../../../assets/css/_radius';
@import '../../../assets/css/_shadow';
#wrap-ratings {
  position: relative;
  .wrap-stars {
    height: 30px;
    float: left;
    position: relative;
    background: #dedede;
  }
  .rating-percent {
    position: absolute;
    content: '';
    top: 0;
    left: 0;
    height: 30px;
    z-index: 1;
    @extend .primary;
  }
  .wrap-stars .one-star {
    height: 30px;
    width: 30px;
    position: relative;
    z-index: 2;
    float: left;
    cursor: pointer;
    background-image: url('../../../assets/images/star.png');
    background-repeat: no-repeat;
    background-position: 0 0;
    background-size: 100% 100%;
  }
  i {
    margin: auto 10px;
    padding: $small/2 $medium;
    width: 30px;
    text-align: center;
    background: #42413e;
    color: #ffffff;
    border-radius: 3px;
    font-style: normal !important;
    float: left;
  }
  .rating-errors {
    position: absolute;
    top: 55px;
    color: $grey;
    right: $small;
    padding: $small $medium;
    display: block;
    background: $white;
    z-index: 1500000000;
    @extend .border, .b-grey, .radius-1, .shadow-2;
  }
  .rating-errors i {
    content: 'Close';
    cursor: pointer;
    text-align: center;
    color: $white !important;
    background: $black;
    font-size: 10px;
    position: absolute;
    top: -20px;
    padding: 7px;
    right: -12px;
    height: 10px;
    @extend .shadow-4;
  }
}
