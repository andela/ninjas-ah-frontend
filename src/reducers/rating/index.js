import { user as initialState } from '../../store/initialState';
import createRatingReducer from './createRatingReducer';

export default (state = initialState, action) => {
  const createRating = createRatingReducer(state, action);
  return createRating || state;
};
