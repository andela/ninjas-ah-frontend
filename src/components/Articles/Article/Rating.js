import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import { createRate, fetchOneArticle } from '../../../actions';
<<<<<<< HEAD
import { clearHighlightArticleStore } from '../../../actions/articles';
=======
>>>>>>> [feature #165412889 && #165412878] add article rating && article read time
import './Rating.scss';

export class Rating extends Component {
  state = { rating: 0, slug: '' };

  componentDidMount() {
    this.setState({ rating: this.props.rating });
  }

  componentWillReceiveProps(nextProps) {
    const { errors, message, article } = nextProps;
    if (message) {
      this.setState({ message });
    }
    if (errors) {
      this.setState({ errors });
    }
    if (Object.keys(article).length !== 0) {
      this.setState({ rating: article.rating });
    }
    this.setState({ loaded: true });
  }

  submitRate = async (rating) => {
<<<<<<< HEAD
    const { isAuth, slug, fetchOneArticle, clearHighlightArticleStore } = this.props;
    if (isAuth) {
      await this.props.createRate(slug, rating);
      await fetchOneArticle(slug);
      clearHighlightArticleStore();
=======
    const { isAuth, slug, fetchOneArticle } = this.props;
    if (isAuth) {
      await this.props.createRate(slug, rating);
      await fetchOneArticle(slug);
>>>>>>> [feature #165412889 && #165412878] add article rating && article read time
    } else {
      this.setState({ errors: { token: 'Failed to authenticate token' } });
    }
  };

  closeMessage = () => {
    this.setState({ message: '' });
    this.setState({ errors: '' });
  };

  ratingStars = () => (
    <div>
      <span className="bold small-padding light left radius-1 medium-h-margin">Rating</span>
      <div className="wrap-stars">
        <div className="rating-percent" style={{ width: `${(this.state.rating * 100) / 5}%` }} />
        {Array.from(Array(5), (e, i) => (
          <div
            onClick={() => this.submitRate(i + 1)}
            key={i + 1}
            className={this.state.rating > i ? 'one-star' : 'one-star'}
          >
            {this.state.rating > i}
          </div>
        ))}
      </div>
      <i>{this.state.rating}</i>
    </div>
  );

  render() {
    const { errors, message } = this.state;
    return (
      <span id="wrap-ratings" className="right">
        {this.ratingStars()}
        {(errors && Object.keys(errors).length && !errors.token)
        || (message && Object.keys(message).length) ? (
          <span className="rating-errors">
            {errors.rating || message}
            <i onClick={() => this.closeMessage()}>Close</i>
          </span>
          ) : (
            ''
          )}
        {errors && errors.token ? (
          <span className="rating-errors">
            Login first to perform this action.{' '}
            <Link className="text-info" to={`/login?redirect=article/${this.props.slug}`}>
              Login
            </Link>
            <i onClick={() => this.closeMessage()}>Close</i>
          </span>
        ) : (
          ''
        )}
      </span>
    );
  }
}

Rating.propTypes = {
  rating: PropTypes.any,
  errors: PropTypes.any,
  slug: PropTypes.string,
  ratingStars: PropTypes.func,
  createRate: PropTypes.func,
<<<<<<< HEAD
  clearHighlightArticleStore: PropTypes.func,
=======
>>>>>>> [feature #165412889 && #165412878] add article rating && article read time
  message: PropTypes.string,
  fetchOneArticle: PropTypes.func.isRequired,
  article: PropTypes.object,
  isAuth: PropTypes.bool
};
export const mapStateToProps = ({
  user: { isAuth },
  articles: { article },
  rating: { createRate: { loading, message, errors } }
}) => ({ loading, message, errors, article, isAuth });

export default connect(
  mapStateToProps,
<<<<<<< HEAD
  { createRate, fetchOneArticle, clearHighlightArticleStore }
=======
  { createRate, fetchOneArticle }
>>>>>>> [feature #165412889 && #165412878] add article rating && article read time
)(Rating);
