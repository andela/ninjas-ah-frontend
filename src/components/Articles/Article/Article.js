import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import 'dotenv/config';
import MetaTags from 'react-meta-tags';
import LazyLoad from 'react-lazyload';
import { Editor, EditorState, convertFromRaw } from 'draft-js';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faClock } from '@fortawesome/free-solid-svg-icons';
import { fetchOneArticle } from '../../../actions';
import avatar from '../../../assets/images/user.png';
import timeStamp from '../../../helpers/timeStamp';
import { NotFound, ProgressBar } from '../../common';
import Layout from '../../Layout';
import './Article.scss';
import Rating from './Rating';
import Comments from '../Comments/Comments';

const { REACT_APP_IMAGE_BASE_URL } = process.env;

export class Article extends Component {
  state = {
    article: {},
    loaded: false,
    imageRectangle:
      'c_fill,g_auto,h_350,w_970/b_rgb:000000,e_gradient_fade,y_-0.20/c_scale,co_rgb:ffffff',
    editorState: EditorState.createEmpty()
  };

  async getSpecificArticle(slug) {
    const { fetchOneArticle } = this.props;
    await fetchOneArticle(slug);

    const { article } = this.props;
    const editorState = article.body
      ? EditorState.createWithContent(convertFromRaw(JSON.parse(article.body)))
      : EditorState.createEmpty();
    this.setState({ loaded: true });
    return this.setState({ article, editorState });
  }

  componentDidMount() {
    const { match: { params: { slug } } } = this.props;
    this.getSpecificArticle(slug);
  }

  render() {
    const { imageRectangle, article, editorState, loaded } = this.state;
    const { loading } = this.props;
    return (
      <Layout>
        <div id="article">
          {loading ? <ProgressBar /> : ''}
          {article && article.id ? (
            <div className="row">
              <MetaTags>
                <title> {article.title || 'Welcome'} - Authors Haven</title>
                <meta
                  name={article.description || 'Authors Haven'}
                  content={
                    article.content
                      ? article.content.replace(/<[^>]*>?/gm, '').trim(60)
                      : 'Authors Haven'
                  }
                />
                <meta property="og:title" content={article.title || 'Authors Haven'} />
                <meta property="og:image" content={article.coverUrl || ''} />
              </MetaTags>
              <LazyLoad height={350}>
                {article.coverUrl ? (
                  <div className="image">
                    <img
                      src={`${REACT_APP_IMAGE_BASE_URL}/${imageRectangle}/${article.coverUrl}`}
                      alt={article.title}
                    />
                  </div>
                ) : (
                  ''
                )}
              </LazyLoad>

              <div className="container">
                <h1 className="xxlarge-text medium-v-padding">{article.title}</h1>
                <div className="articleInfo">
                  <div className="row">
                    <div className="small-screen-4 medium-screen-2 large-screen-2 large-v-padding">
                      <span className="avatar">
                        <img src={avatar} alt={article.title} />
                        <span>{' John Doe'}</span>
                      </span>
                      <Link to="/" className="button small-button primary">
                        Follow
                      </Link>
                      <span className="medium-h-padding">{timeStamp(article.createdAt)}</span>
                      <span className="medium-h-padding">
                        <FontAwesomeIcon icon={faClock} className="text-light-grey" />{' '}
                        {article.readTime} min read
                      </span>
                    </div>
                    <div className="small-screen-4 medium-screen-2 large-screen-2">
                      <Rating slug={article.slug} rating={article.rating || 0} />
                    </div>
                  </div>
                </div>
                <br />
                <div className="large-text">{article.description}</div>

                <div className="divider light" />
                <div className="articleBody">
                  {article && <Editor editorState={editorState} readOnly={false} />}
                </div>
                <Comments slug={article.slug} />
              </div>
            </div>
          ) : (
            <div>{loaded && !Object.keys(article).length ? <NotFound /> : <div />}</div>
          )}
        </div>
      </Layout>
    );
  }
}

Article.defaultProps = { match: { params: { slug: '' } } };

Article.propTypes = {
  article: PropTypes.object,
  fetchOneArticle: PropTypes.func.isRequired,
  editorState: PropTypes.func,
  match: PropTypes.object,
  slug: PropTypes.string,
  rating: PropTypes.number,
  params: PropTypes.object,
  message: PropTypes.object,
  errors: PropTypes.object,
  loaded: PropTypes.bool,
  loading: PropTypes.bool
};
const mapStateToProps = ({ articles: { article, errors, loading } }) => ({
  article,
  errors,
  loading
});

export default connect(
  mapStateToProps,
  { fetchOneArticle }
)(Article);
