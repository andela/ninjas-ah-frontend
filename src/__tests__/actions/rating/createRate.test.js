import React from 'react';
import configureMockStore from 'redux-mock-store';
import thunk from 'react-thunk';
import { Provider } from 'react-redux';
import { createRate } from '../../../actions/rating';
import { Rating as RatingComponent } from '../../../components/Articles/Article/Rating';
import { shallow, mount } from '../../../../config/enzymeConfig';

describe('<RatingComponent />', () => {
  const props = {
    slug: 'slug-slug',
    rating: 1,
    errors: { error: ['12'] },
    createRate: jest.fn(),
    fetchOneArticle: jest.fn()
  };
  const rating = {
    slug: 'slug-slug',
    rating: 1
  };
  const component = shallow(<RatingComponent {...props} />);
  it('should render a <CreateRatingComponent /> component ', () => {
    expect(component).toMatchSnapshot();
  });
  const dispatch = jest.fn(action => action);

  describe('Create rating', () => {
    test('returns rating information', async () => {
      const result = createRate(rating)(dispatch);
      expect(result).toHaveProperty('type');
      expect(result).toHaveProperty('payload');
    });
  });
});
