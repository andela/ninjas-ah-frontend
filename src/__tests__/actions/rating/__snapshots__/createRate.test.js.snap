// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RatingComponent /> should render a <CreateRatingComponent /> component  1`] = `ShallowWrapper {}`;
